src/storage/mvcc/txn.rs

```diff
@@ -856,29 +844,44 @@ mod tests {
         must_get(&engine, k, t4, v);
     }
 
-    #[test]
-    fn test_mvcc_txn_rollback() {
-        test_mvcc_txn_rollback_imp(b"k", b"v");
-
-        let long_value = "v".repeat(SHORT_VALUE_MAX_LEN + 1).into_bytes();
-        test_mvcc_txn_rollback_imp(b"k2", &long_value);
-    }
-
-    fn test_mvcc_txn_rollback_err_imp(k: &[u8], v: &[u8]) {
+    fn test_mvcc_txn_rollback_imp(k: &[u8], v: &[u8]) {
         let engine = TestEngineBuilder::new().build().unwrap();
 
         must_prewrite_put(&engine, k, v, k, 5);
-        must_commit(&engine, k, 5, 10);
-        must_rollback_err(&engine, k, 5);
-        must_rollback_err(&engine, k, 5);
+        must_rollback(&engine, k, 5);
+        // Rollback should be idempotent
+        must_rollback(&engine, k, 5);
+        // Lock should be released after rollback
+        must_unlocked(&engine, k);
+        must_prewrite_lock(&engine, k, k, 10);
+        must_rollback(&engine, k, 10);
+        // data should be dropped after rollback
+        must_get_none(&engine, k, 20);
+
+        // Can't rollback committed transaction.
+        must_prewrite_put(&engine, k, v, k, 25);
+        must_commit(&engine, k, 25, 30);
+        must_rollback_err(&engine, k, 25);
+        must_rollback_err(&engine, k, 25);
```

Comment:
What's for a repeated check?

Reply:
Maybe trying to make sure the commit message is not removed, so the second rollback will still fail? I don't know. The old code did this too.